server/src/apub/fetcher.rs

```diff
@@ -43,36 +38,111 @@ use chrono::NaiveDateTime;
   let well_known_uri = Url::parse(&format!(
     "{}://{}/.well-known/nodeinfo",
     get_apub_protocol_string(),
     domain
   ))?;
-  let well_known = fetch_remote_object::<NodeInfoWellKnown>(&well_known_uri)?;
-  Ok(fetch_remote_object::<NodeInfo>(&well_known.links.href)?)
+
+  let well_known = fetch_remote_object::<NodeInfoWellKnown>(client, &well_known_uri).await?;
+  let nodeinfo = fetch_remote_object::<NodeInfo>(client, &well_known.links.href).await?;
+
+  Ok(nodeinfo)
+}
+
+enum RequestError {
+  Send(actix_web::client::SendRequestError),
+  Recv(awc::error::JsonPayloadError),
+}
+
+impl RequestError {
+  fn is_connect_timeout(&self) -> bool {
+    if let RequestError::Send(e) = self {
+      if let actix_web::client::SendRequestError::Connect(e) = e {
+        if let actix_web::client::ConnectError::Timeout = e {
+          return true;
+        }
+      }
+    }
+
+    false
+  }
+
+  fn into_lemmy_error(self) -> LemmyError {
+    match self {
+      RequestError::Send(e) => SendError(e.to_string()).into(),
+      RequestError::Recv(e) => RecvError(e.to_string()).into(),
+    }
+  }
+}
+
+async fn do_fetch<Response>(
+  client: &Client,
+  url: &Url,
+  timeout: Duration,
+) -> Result<Response, RequestError>
+where
+  Response: for<'de> Deserialize<'de>,
+{
+  let json = client
+    .get(url.as_str())
+    .header("Accept", APUB_JSON_CONTENT_TYPE)
+    .timeout(timeout)
+    .send()
+    .await
+    .map_err(|e| {
+      debug!("Send error, {}", e);
+      RequestError::Send(e)
+    })?
+    .json()
+    .await
+    .map_err(|e| {
+      debug!("Receive error, {}", e);
+      RequestError::Recv(e)
+    })?;
+
+  Ok(json)
 }
 
 /// Fetch any type of ActivityPub object, handling things like HTTP headers, deserialisation,
 /// timeouts etc.
-pub fn fetch_remote_object<Response>(url: &Url) -> Result<Response, Error>
+pub async fn fetch_remote_object<Response>(
+  client: &Client,
+  url: &Url,
+) -> Result<Response, LemmyError>
 where
   Response: for<'de> Deserialize<'de>,
 {
   if !is_apub_id_valid(&url) {
-    return Err(format_err!("Activitypub uri invalid or blocked: {}", url));
+    return Err(format_err!("Activitypub uri invalid or blocked: {}", url).into());
   }
-  // TODO: this function should return a future
+
+  let mut err = format_err!("This should never happen").into();
+
   let timeout = Duration::from_secs(60);
-  let text = Request::get(url.as_str())
-    .header("Accept", APUB_JSON_CONTENT_TYPE)
-    .connect_timeout(timeout)
-    .timeout(timeout)
-    .body(())?
-    .send()?
-    .text()?;
-  let res: Response = serde_json::from_str(&text)?;
-  Ok(res)
+  let start = Instant::now();
+
+  // retry up to 2 additional times as long as we haven't hit the timeout
+  for i in 0..3 {
+    match do_fetch(client, url, timeout).await {
+      Ok(json) => return Ok(json),
+      Err(e) if e.is_connect_timeout() => {
+        err = e.into_lemmy_error();
+      }
+      Err(e) => return Err(e.into_lemmy_error()),
+    }
```

Comment:
Return immediately for success and most errors, but continue on if the error is a connection timeout, since this is caused by a failure in TRust DNS Resolver (sometimes).

Reply:
I have been occasionally running into this when testing. When I retry the request it usually works fine.

```
lemmy-beta_1      | [2020-06-29T14:42:12Z DEBUG lemmy_server::apub] Checking http://lemmy-alpha:8540/inbox
lemmy-beta_1      | [2020-06-29T14:42:12Z DEBUG trust_dns_resolver::async_resolver] trust-dns resolver running
lemmy-beta_1      | [2020-06-29T14:42:14Z DEBUG trust_dns_proto::xfer::dns_handle] querying: lemmy-alpha A
lemmy-beta_1      | [2020-06-29T14:42:14Z DEBUG trust_dns_resolver::name_server::name_server_pool] sending request: [Query { name: Name { is_fqdn: false, labels: [lemmy-alpha] }, query_type: A, query_class: IN }]
lemmy-beta_1      | [2020-06-29T14:42:14Z DEBUG trust_dns_resolver::name_server::name_server] reconnecting: NameServerConfig { socket_addr: V4(127.0.0.11:53), protocol: Udp, tls_dns_name: None }
lemmy-beta_1      | [2020-06-29T14:42:14Z DEBUG trust_dns_proto::xfer] enqueueing message: [Query { name: Name { is_fqdn: false, labels: [lemmy-alpha] }, query_type: A, query_class: IN }]
lemmy-beta_1      | [2020-06-29T14:42:14Z DEBUG trust_dns_proto::udp::udp_stream] created socket successfully
lemmy-beta_1      | [2020-06-29T14:42:14Z ERROR lemmy_server::websocket::server] Error during message handling Error sending request, Failed to connect to host: Timeout out while establishing connection
```
You said its not an issue in the release builds?

Reply:
that is correct. Do you know which request that is? it may need the same retry logic from fetcher

Reply:
actually it would be safe to assume all actix-web client requests should get the retry logic to make dealing with debug builds less annoying

Reply:
It was the one in `activities.rs`: `send_activity`

Reply:
I've added retry to every http request I found